File: top_interview_questions/1_easy/1_array/intersection_of_two_arrays_ii/solution.py
import collections
from typing import List


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        nums1_counter = collections.Counter(nums1)
        nums2_counter = collections.Counter(nums2)

        result = []
        for num in nums1_counter:
            if num in nums2_counter:
                min_count = min(nums1_counter[num], nums2_counter[num])
                result.extend([num]*min_count)

        return result

File: top_interview_questions/1_easy/1_array/intersection_of_two_arrays_ii/solution.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(intersect([]int{1, 2, 2, 1}, []int{2, 2})) // expected []int{2, 2}
}

func intersect(nums1 []int, nums2 []int) []int {
	var result []int
	counter1 := counter(nums1)
	counter2 := counter(nums2)

	for num, count := range counter1 {
		minCount := min(count, counter2[num])
		for i := 0; i < minCount; i++ {
			result = append(result, num)
		}
	}

	return result
}

func counter(nums []int) map[int]int {
	c := make(map[int]int)

	for _, num := range nums {
		c[num]++
	}

	return c
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
